### DISCUSSION TITLE
Questions about the provided examples

### DISCUSSION THREAD

**Discussion by KuanYuChang** [cda-tum/mqt-qmap#1] [General]
Hi there,

It seems that the examples provided in this repo are decomposed into elementary operations and written in OpenQASM format. Is there any instructions for preparing decomposed quantum circuits or even for the other format, e.g., Real?

Sincerely.

**Comment by burgholzer** [comment]
Hi,

That is correct. Our tool assumes that the circuit to be mapped has already been composed into a sequence of elementary gates (i.e., single- or two-qubit gates) supported by the device it shall be mapped to.
At the moment, we have _not_ integrated any approaches for carrying out this *synthesis* or *decomposition* step into the JKQ QMAP tool.
However, you could, for example, use IBM's Qiskit to generate such decomposed circuits and dump them to `.qasm` files.

As for `.real` files: Since these typically realize reversible functionality, they rely mainly on multi-controlled Toffoli and Fredkin gates.
Consequently, a straightforward synthesis approach would be to sequentially apply any known decomposition of these gates into elementary gates.

In case of further questions, do not hesitate to ask!

Best regards,
Lukas

**Comment by KuanYuChang** [comment]
Hi Lukas, 

Thanks for your quick reply, I will take my time to look into the transpiling process provided by IBM's Qiskit.

Sincerely,
Kuan-Yu

**Comment by burgholzer** [comment]
You are welcome! 
I'll close this issue for now. In case of further questions do not hesitate to ask.

Best regards,
Lukas

**Comment by KuanYuChang** [comment]
Hi Lukas,

Another question about decomposing quantum circuits using basis gates, `['x', 'h', 't', 'tdg', 'cx']`, is there any tips or suggestions to work around it? 

Take `alu-v4_36` for example, the circuit could be represented in MCT-Library as following:

```
OPENQASM 2.0;
include "qelib1.inc";
qreg q[5];
mcx q[4],q[3],q[2],q[1];
cx q[3],q[4];
ccx q[1],q[3],q[2];
x q[2];
ccx q[4],q[2],q[0];
ccx q[0],q[4],q[2];
cx q[2],q[4];
```

When I tried to decompose `alu-v4_36` using Qiskit, it says:
> "Cannot unroll the circuit to the given basis, ['x', 'h', 't', 'tdg', 'cx']. No rule to expand instruction u1."

However, the same circuit provided in examples of this repo decomposed only based on `['x', 'h', 't', 'tdg', 'cx']`.

Sincerely,
Kuan-Yu

---

The following provides additional information:
- `alu-v4_36` in this repo: https://github.com/iic-jku/qmap/blob/master/examples/alu-v4_36.qasm
- `alu-v4_36` from RevLib: http://www.revlib.org/doc/real/alu-v4_36.real
- Version of Qiskit: 0.19.0
- Script to reproduce:
``` Python
import qiskit as qk
circ = qk.QuantumCircuit(5)
circ.mcx([4, 3, 2], 1)
circ.cx(3, 4)
circ.ccx(1, 3, 2)
circ.x(2)
circ.ccx(4, 2, 0)
circ.ccx(0, 4, 2)
circ.cx(2, 4)
circ = qk.transpile(circ, basis_gates=['x', 'h', 't', 'tdg', 'cx'])
```
- Detail traceback:
```
Traceback (most recent call last):
  File "produce.py", line 14, in <module>
    circ = qk.transpile(circ, basis_gates=['x', 'h', 't', 'tdg', 'cx'])
  File "${LIB}/python3.6/site-packages/qiskit/compiler/transpile.py", line 210, in transpile
    circuits = parallel_map(_transpile_circuit, list(zip(circuits, transpile_args)))
  File "${LIB}/python3.6/site-packages/qiskit/tools/parallel.py", line 105, in parallel_map
    return [task(values[0], *task_args, **task_kwargs)]
  File "${LIB}/python3.6/site-packages/qiskit/compiler/transpile.py", line 307, in _transpile_circuit
    output_name=transpile_config['output_name'])
  File "${LIB}/python3.6/site-packages/qiskit/transpiler/passmanager.py", line 214, in run
    return self._run_single_circuit(circuits, output_name, callback)
  File "${LIB}/python3.6/site-packages/qiskit/transpiler/passmanager.py", line 277, in _run_single_circuit
    result = running_passmanager.run(circuit, output_name=output_name, callback=callback)
  File "${LIB}/python3.6/site-packages/qiskit/transpiler/runningpassmanager.py", line 115, in run
    dag = self._do_pass(pass_, dag, passset.options)
  File "${LIB}/python3.6/site-packages/qiskit/transpiler/runningpassmanager.py", line 145, in _do_pass
    dag = self._run_this_pass(pass_, dag)
  File "${LIB}/python3.6/site-packages/qiskit/transpiler/runningpassmanager.py", line 157, in _run_this_pass
    new_dag = pass_.run(dag)
  File "${LIB}/python3.6/site-packages/qiskit/transpiler/passes/basis/unroller.py", line 107, in run
    unrolled_dag = self.run(decomposition)  # recursively unroll ops
  File "${LIB}/python3.6/site-packages/qiskit/transpiler/passes/basis/unroller.py", line 107, in run
    unrolled_dag = self.run(decomposition)  # recursively unroll ops
  File "${LIB}/python3.6/site-packages/qiskit/transpiler/passes/basis/unroller.py", line 107, in run
    unrolled_dag = self.run(decomposition)  # recursively unroll ops
  File "${LIB}/python3.6/site-packages/qiskit/transpiler/passes/basis/unroller.py", line 93, in run
    (str(self.basis), node.op.name))
qiskit.exceptions.QiskitError: "Cannot unroll the circuit to the given basis, ['x', 'h', 't', 'tdg', 'cx']. No rule to expand instruction u1."
```

**Comment by burgholzer** [comment]
Hi Kuan-Yu,

Qiskit (or in general IBM devices) support the universal gate set `['CX', 'U']` where `U(theta, phi,lambda)` represents arbitrary single-qubit gates. All of the gates that Qiskit understands are derived from this gate-set. You can find these definitions at https://github.com/Qiskit/qiskit-terra/blob/master/qiskit/qasm/libs/qelib1.inc
There you can see that `t = u1(pi/4)`, `tdg = u1(-pi/4)` and `h = u2(0, pi)`.

To fix your error just pass `basis_gates=['u3', 'cx']` (or anything more specific, like ``basis_gates=['x', 'h', 't', 'tdg', 'u1', 'u2', 'u3', 'cx']``) to the transpile function.

Best regards,
Lukas

**Comment by KuanYuChang** [comment]
Hi Lukas,

Following your suggestion, I have done the decomposition of `alu-v4_36` by passing an additional gate `u1`, i.e., passing `basis_gates=['cx', 'x', 'h', 't', 'tdg', 'u1']`, but I'm still wondering how to done it without using `u1`, i.e., used only `['cx', 'x', 'h', 't', 'tdg']`.

Since I only found the reference that decomposing the MCT gate based on NCV-Library, could you provide some reference to decompose the MCT gate based on `['cx', 'x', 'h', 't', 'tdg']`?

Sincerely,
Kuan-Yu

**Comment by burgholzer** [comment]
Hi Kuan-Yu,

In fact `u1(pi/4)` is equivalent to `t`. That is just how Qiskit breaks down the operations. There is not really a reason for not including this gate. 
Our tool for example will read in the `.qasm` file and automatically convert `u1(pi/4)` to `t`.

The standard reference for this kind of topic would probably be https://arxiv.org/abs/quant-ph/9503016.
Qiskit implements many of the techniques described in https://arxiv.org/abs/1508.03273.
You can look at the source code directly at https://qiskit.org/documentation/_modules/qiskit/circuit/quantumcircuit.html#QuantumCircuit.mcx.

Hope this helps,
Lukas

**Comment by KuanYuChang** [comment]
Hi Lukas,

I truly appreciate your support, but it seems that the provided reference will take me a long time to understand. I think I will use `basis_gates=['cx', 'u3']` for the further circuits for now.

Many thanks,
Kuan-Yu
